dragonfly/block/fire.go

```diff
@@ -0,0 +1,257 @@
+}
+
+// burn attempts to burn a block.
+func (f Fire) burn(pos world.BlockPos, w *world.World, chanceBound int) {
+	if flammable, ok := w.Block(pos).(Flammable); ok && rand.Intn(chanceBound) < flammable.FlammabilityInfo().Flammability {
+		//TODO: Check if not raining
+		if rand.Intn(f.Age+10) < 5 {
+			age := f.Age + rand.Intn(5)/4
+			if age > 15 {
+				age = 15
+			}
+			w.PlaceBlock(pos, Fire{Type: f.Type, Age: age})
+		} else {
+			w.BreakBlockWithoutParticles(pos)
+		}
+		//TODO: Light TNT
+	}
+}
+
+// tick ...
+func (f Fire) tick(pos world.BlockPos, w *world.World) {
+	if f.Type == fire.Normal() {
+		infinitelyBurns := infinitelyBurning(pos, w)
+
+		// TODO: !infinitelyBurning && raining && exposed to rain && 20 + age * 3% = extinguish & return
+
+		if f.Age < 15 && rand.Intn(3) == 0 {
+			f.Age++
+			w.PlaceBlock(pos, f)
+		}
+
+		w.ScheduleBlockUpdate(pos, time.Duration(30+rand.Intn(10))*time.Second/20)
+
+		if !infinitelyBurns {
+			if !neighboursFlammable(pos, w) {
+				if !w.Block(pos.Side(world.FaceDown)).Model().FaceSolid(pos, world.FaceUp, w) || f.Age > 3 {
+					w.BreakBlockWithoutParticles(pos)
+				}
+				return
+			}
+			if !FlammableBlock(w.Block(pos.Side(world.FaceDown))) && f.Age == 15 && rand.Intn(4) == 0 {
+				w.BreakBlockWithoutParticles(pos)
+				return
+			}
+		}
+
+		//TODO: If high humidity, chance should be subtracted by 50
+		for face := world.Face(0); face < 6; face++ {
+			if face == world.FaceUp || face == world.FaceDown {
+				f.burn(pos.Side(face), w, 300)
+			} else {
+				f.burn(pos.Side(face), w, 250)
+			}
+		}
+
+		for y := -1; y <= 4; y++ {
+			randomBound := 100
+			if y > 1 {
+				randomBound += (y - 1) * 100
+			}
+
+			for x := -1; x <= 1; x++ {
+				for z := -1; z <= 1; z++ {
+					if x != 0 || y != 0 || z != 0 {
+						blockPos := pos.Add(world.BlockPos{x, y, z})
+						block := w.Block(blockPos)
+						if _, ok := block.(Air); !ok {
+							continue
+						}
+
+						encouragement := 0
+						blockPos.Neighbours(func(neighbour world.BlockPos) {
+							if flammable, ok := w.Block(neighbour).(Flammable); ok {
+								encouragement = max(encouragement, flammable.FlammabilityInfo().Encouragement)
+							}
+						})
+						if encouragement <= 0 {
+							continue
+						}
+
+						//TODO: Divide chance by 2 in high humidity
+						maxChance := (encouragement + 40 + difficultyOffset(w.Difficulty())) / (f.Age + 30)
+
+						//TODO: Check if exposed to rain
+						if maxChance > 0 && rand.Intn(randomBound) <= maxChance {
+							age := f.Age + rand.Intn(5)/4
+							if age > 15 {
+								age = 15
+							}
+							w.PlaceBlock(blockPos, Fire{Type: f.Type, Age: age})
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+// EntityCollide ...
+func (f Fire) EntityCollide(e world.Entity) {
+	if flammable, ok := e.(entity.Flammable); ok {
+		flammable.FireDamage(1)
+		flammable.SetOnFire(160)
+	}
+}
+
+// ScheduledTick ...
+func (f Fire) ScheduledTick(pos world.BlockPos, w *world.World) {
+	f.tick(pos, w)
```

Comment:
This will make fire spread very quickly if you spam place blocks next to it. Is that intended?